### PR TITLE
fix(mcp): annotate connected-app reachability on the gateway connect page

### PR DESCRIPTION
**Pull request by tin-berri**
## TLDR

Problem this solves:

- Connect page shows servers a gateway session is never served
- Real report: page shows 7 servers / 253 tools, client gets 171
- Nothing on the page marks which servers actually count

How it solves it:

- GET /v1/mcp/server?connected_app_view=true stamps per-server reachability
- Computed by the exact resolver pair the live session uses
- Connect page shows only servers the session will be served

## Relevant issues

- The MCP connect page resolves its grid through the dashboard identity, so an admin sees the whole registry, while the gateway DCR session minted by that page resolves servers as an admitted subject through grant sources only
- This PR adds an opt-in annotation to the server list endpoint plus connect-mode rendering that hides servers the session will not be served, so the page and the session can no longer disagree
- Existing consumers of GET /v1/mcp/server are unchanged; the field stays null unless the new param is passed

## Linear ticket

Resolves LIT-4861

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [ ] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Screenshots / Proof of Fix

Rig: proxy on localhost:4861 backed by real Postgres, four streamable HTTP MCP upstreams serving three tools each, and four servers registered through the API to exercise every grant channel: u1srv granted through the signed-in user's own object permission, t1srv granted through a roster team, a1srv flagged allow_all_keys, x1srv with no grant. The signed-in user is a proxy admin, matching the reported setup. The client leg is the full gateway DCR flow over the wire: /register, /authorize (303 to /ui/chat/integrations?connect_flow=...), POST /authorize/complete with the session cookie, /token with PKCE, then MCP initialize and tools/list with the minted session bearer

Before, at base bdf8f8c309:

```
$ curl -s "http://localhost:4861/v1/mcp/server" -H "Authorization: Bearer $UI_SESSION_KEY"
servers listed: 4        # a1srv, t1srv, u1srv, x1srv; no reachability field exists

$ curl -s "http://localhost:4861/v1/mcp/server?connected_app_view=true" -H "Authorization: Bearer $UI_SESSION_KEY"
servers listed: 4 | any connected_app_reachable field: False        # unknown param silently dropped

$ ./dcr_flow.sh          # register -> authorize -> complete -> token -> MCP tools/list
served tool count: 9
served server prefixes: ['a1srv', 't1srv', 'u1srv']        # x1srv absent, no error anywhere
```

The page lists 4 connectable servers while the session it just created is served 3. This is the reported 253 vs 171 mismatch reduced to a rig

After, at e2444030e8, same commands against the same database:

```
$ curl -s "http://localhost:4861/v1/mcp/server?connected_app_view=true" -H "Authorization: Bearer $UI_SESSION_KEY"
  a1srv | connected_app_reachable: True
  t1srv | connected_app_reachable: True
  u1srv | connected_app_reachable: True
  x1srv | connected_app_reachable: False

$ curl -s "http://localhost:4861/v1/mcp/server" -H "Authorization: Bearer $UI_SESSION_KEY"
servers listed: 4 | flags: ['None']        # default unchanged without the param

$ ./dcr_flow.sh
served tool count: 9
served server prefixes: ['a1srv', 't1srv', 'u1srv']        # equals the flagged-true set exactly
```

UI verification steps: run the dashboard dev server, start a gateway connect flow from any MCP client so the browser lands on /ui/chat/integrations?connect_flow=..., and observe that servers without a grant do not appear at all, while the plain Integrations page without connect_flow renders exactly as before

## Type

🐛 Bug Fix

## Changes

- GET /v1/mcp/server gains an opt-in connected_app_view query param; response rows gain connected_app_reachable, default null
- A new helper builds the admitted-subject auth with MCPRequestHandler._reload_admitted_user and resolves the server set with MCPServerManager.get_allowed_mcp_servers, the same two owners the live /mcp session uses, so the page and the session cannot drift
- Any failure to build the admitted set marks every server unreachable; a user whose reload fails cannot open a gateway session either, so this is the true answer rather than a fallback
- The stamp lands after both list branches, so view_all deployments (whose management-endpoint shortcut preempts the server manager's admin shortcut) and team_id queries are annotated uniformly, and it lands before the sanitizers, which preserve it
- MCPAppsPanel in connect mode requests the flag and hides unreachable servers entirely; the filtered list is the single source for counts, tabs, auto-select, detail view, and tool-count fetches. Client-forwarded auth types keep their existing "Not supported on this connection" label (a server property, unlike reachability, which is caller-relative), and a user with zero reachable servers gets an explanatory empty state

Two things a reviewer may ask. The API annotates rather than filters so the reachability field stays caller-relative and every other consumer of GET /v1/mcp/server is untouched; the connect page hides the annotated-unreachable rows client side, so the page never advertises a server the session cannot use, and the annotation still lands uniformly on the view_all and team_id branches. build_effective_auth_contexts now appends the admitted-user context for UI session credentials, so the page list, tool counts, Connect actions, and credential endpoints all resolve every server the user is granted through any channel identically; the page's membership equals the admitted set by construction and no row can render whose follow-up actions would be denied. On the REST tool routes a non-admin dashboard session is swapped for the admitted-subject principal at the route boundary (acting_user_auth), so tool permission ceilings, reachability, rate limits, and billing bind exactly as they do for a gateway session; admin sessions keep their operator view and caller-passed keys are never widened. The view is honored only for the dashboard's UI session credential, via is_ui_session_credential factored out of resolve_ui_session_team_ids so both user-identity widening sites share one trust boundary; a caller-passed virtual key gets the param as a no-op, so a scoped key can never enumerate its owning user's wider grants The keyless /ui/connect page (which renders the same panel without connect mode) and the enable_chat_ui gating of /ui/chat/integrations are pre-existing gaps tracked on the Linear ticket, kept out of this PR's scope

Three things changed after the last review round, all at the class rather than at the reported call

The connect page's detail view no longer keeps its own copy of the server object. It stores the selected server's id and derives the row from the list, so a server a refetch drops as unreachable cannot be the detail view's subject at all; the stale render, the stale tools query and the Connect action that fell through a missing-target lookup stop being reachable states instead of being blocked one at a time. handleToggle takes the server it is toggling, which deletes the lookup that could miss. Pinned by a test that opens the detail view, refetches a list without that server, and asserts the view falls back to the grid with no connect action anywhere; storing the object again fails it

admitted_user_context is now the single owner of "what user identity does this dashboard session act as". The management endpoint's private copy of the reload, the HTTPException swallow and the logging is gone, along with its dead empty-user_id branch, so the page annotation, build_effective_auth_contexts and the tool routes cannot answer that question differently. The owner also carries the request's tracing span onto the admitted principal, because _reload_admitted_user builds a fresh auth from the user row and has none of its own, so the previous round's swap had silently detached every downstream lookup and the tool-call logging from the request's trace

Toolset scoping and the acting-as-user swap are mutually exclusive, and they now share one owner on the tools list route. The admitted subject resolves per grant source and a team source deliberately carries none of the caller's object_permission, so a toolset narrowing layered on top would evaporate on every team-granted server: a non-admin dashboard user whose own object permission grants a toolset would be admitted through that grant and then served tools from servers the toolset never named. A request carrying a toolset name keeps the caller's own credential, exactly as it did before the swap. The equivalent composition on /toolset/{name}/mcp for a gateway session predates this PR, is untouched by it, and is tracked separately

A fourth followed from reading the rest of the panel's async surface. Three continuations decided against state captured before their await and committed after it, so a reachability refetch landing in between could not be seen. handleToggle validated the server at click time and then wrote its name into the selection whatever the list had since become, so a server the refresh had dropped was selected anyway; it now re-asks connectableNow at the commit, and that predicate resolves the id against the current list, so absence fails closed rather than reading as nothing to block. The load pipeline was the worse of the two because its cancel flag was shared across runs: the successor's effect body reset it to false before the predecessor's fetch resolved, so a superseded load could still run setServers and put the dropped server back on the page outright. That flag is now a per-effect local only its own cleanup can clear, which is also what makes unmount stop the chunked tool-count loop again, and the load passes its own liveness check down to the tool-count and oauth-status writes instead of having them consult a flag every other run shares. Both are pinned by tests that hold the in-flight request open across the refetch, and each mutation kills exactly its own test

## QA runbook

Against the same rig, with the dashboard dev server on :3000 and the proxy on :4861

1. Sign in as the non-admin user and start a gateway connect flow so the browser lands on /ui/chat/integrations?connect_flow=..., then open any server's detail view and leave it open. In another tab revoke that user's grant for that server, then force the panel to reload its list (sign the dashboard session out and back in, which changes the access token the panel keys its fetch on). The detail view should fall back to the grid with the revoked server gone, and no Connect control for it anywhere
2. `curl -s "http://localhost:4861/mcp-rest/tools/list?toolset_name=<a toolset the user's own object permission does not grant>" -H "Authorization: Bearer $UI_SESSION_KEY"` should still be refused, the same as before this PR; the toolset request must not be admitted through the user's grant union
3. `curl -s "http://localhost:4861/v1/mcp/server?connected_app_view=true" -H "Authorization: Bearer $UI_SESSION_KEY"` should return the same connected_app_reachable flags as before, and the served tool count over the real DCR flow should still equal the flagged-true set exactly

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

---

> [!NOTE]
> **Medium Risk**
> Touches MCP authorization identity resolution (dashboard vs admitted subject) and server listing semantics; behavior is gated to UI session credentials and connect mode, with broad test coverage, but mistakes could affect tool access or visibility for dashboard users.
> 
> **Overview**
> Fixes the gateway **connect** flow showing MCP servers and tool counts that the minted OAuth session never receives.
> 
> **Backend:** `GET /v1/mcp/server` accepts optional `connected_app_view` and sets `connected_app_reachable` on each row using the same admitted-subject resolver as live `/mcp` sessions (UI session credentials only; virtual keys are not widened). `build_effective_auth_contexts` now also includes the admitted user context so listings match user-level grants. MCP REST tool list/call routes resolve non-admin dashboard sessions via `acting_user_auth` (renamed from toolset-only scope), while **toolset-scoped** requests keep the caller credential so narrowing is not lost on team grants.
> 
> **Dashboard:** In connect mode, `MCPAppsPanel` requests the new view, filters out unreachable servers, and hardens async behavior (detail view by server id, per-load cancellation, re-check after Connect) so refetches cannot leave stale UI or selections.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a6d3b1763ca23471e00bef85c0cebccfbd54835d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by tin-berri** [comment]
Addressed the P1 in b6441409b5 at the mechanism level. The unavailability decision now lives in one predicate, connectUnavailabilityLabel, and every surface that could add or count a server consumes it: the card indicator, the detail view action area (which renders the label instead of any Connect control), the toggle-on path in handleToggle, the oauth auto-select effect, and the Connected count, so no interaction path can disagree with the label. This also closed the same pre-existing hole for servers labeled not supported on this connection, whose detail view likewise offered Connect. Regression tests drive the detail view for both unavailable states in connect mode and assert the Connect action is absent, plus pin that it still renders outside connect mode; the detail-view gate is mutation-checked. The eslint suppressions baseline ratchets down by one grandfathered nested ternary removed in the restructure.

@greptileai

**Comment by tin-berri** [comment]
bugbot run

**Comment by tin-berri** [comment]
@greptileai rereview and score properly

**Comment by tin-berri** [comment]
Product decision after review: in connect mode the page now hides annotated-unreachable servers entirely instead of dimming them (ffc22f22e5), so the connect surface lists exactly what the session will be served. The API contract is unchanged (opt-in connected_app_view param, connected_app_reachable field); the filter is applied once where the list is set, so counts, tabs, auto-select, the detail view, and tool-count fetches all derive from the filtered state. Unsupported auth types keep their dimmed label since that is a property of the server rather than of the caller, and a user with zero reachable servers gets an explanatory empty state. Tests updated: hiding is asserted in connect mode, full rendering asserted outside it, and the filter is mutation-checked

@greptileai

**Comment by tin-berri** [comment]
One more commit, 5ace001672: the connect view now guarantees the invariant in both directions. Previously membership came from the dashboard resolver with reachability only annotated on top, so a server the session is served but the dashboard resolver missed would be invisible on the page (an under-report, the mirror image of the reported bug). The connect view now unions in session-reachable servers missing from the dashboard list, built from the registry and passed through the same redaction and sanitizer ladder, so page membership equals the admitted set by construction. Pinned by a test running as an internal user through the non-admin sanitizer, with the union arm mutation-checked

@greptileai

File: litellm/proxy/management_endpoints/mcp_management_endpoints.py [hunk 1]
```diff
@@ -1009,6 +1031,18 @@ async def fetch_all_mcp_servers(
             servers = await _resolve_accessible_mcp_servers(user_api_key_dict)
             redacted_mcp_servers = _redact_mcp_credentials_list(servers)
 
+        if connected_app_view is True:
+            reachable_ids = await _connected_app_reachable_server_ids(user_api_key_dict.user_id)
+            listed_ids = {server.server_id for server in redacted_mcp_servers}
+            missing_tables = [
+                global_mcp_server_manager._build_mcp_server_table(registry_server)
+                for missing_id in sorted(reachable_ids - listed_ids)
```
[the remark this answers is not in the document]
Comment (tin-berri) [reply litellm/proxy/management_endpoints/mcp_management_endpoints.py]: Agreed, fixed in dcfc10d247 with the first option, at the credential boundary rather than per response. The view is now gated on is_ui_session_credential, factored out of resolve_ui_session_team_ids so both user-identity widening sites (this view and build_effective_auth_contexts's team resolution) share the one existing trust boundary: the SSO-minted dashboard session token acting as its user. Any caller-passed virtual key, scoped or not, gets connected_app_view as a no-op and the admitted resolver is never consulted for it, so there is nothing to intersect and no partial mode to reason about. Pinned by two tests: a caller-passed key with the param gets every field unset and the admitted reload is never awaited, and the gate is mutation-checked (removing the credential check fails both)

**Comment by tin-berri** [comment]
@greptileai

**Comment by tin-berri** [comment]
bugbot run

File: litellm/proxy/management_endpoints/mcp_management_endpoints.py [hunk 2]
```diff
@@ -1009,6 +1032,18 @@ async def fetch_all_mcp_servers(
...
 
+        if connected_app_view is True and is_ui_session_credential(user_api_key_dict):
+            reachable_ids = await _connected_app_reachable_server_ids(user_api_key_dict.user_id)
+            listed_ids = {server.server_id for server in redacted_mcp_servers}
+            missing_tables = [
+                global_mcp_server_manager._build_mcp_server_table(registry_server)
+                for missing_id in sorted(reachable_ids - listed_ids)
+                if (registry_server := global_mcp_server_manager.get_mcp_server_by_id(missing_id)) is not None
+            ]
+            redacted_mcp_servers.extend(_redact_mcp_credentials_list(missing_tables))
```
[the remark this answers is not in the document]
Comment (tin-berri) [reply litellm/proxy/management_endpoints/mcp_management_endpoints.py]: Correct finding, and it pointed at the real root: build_effective_auth_contexts carries team contexts but never the user row's own object permission, so the list endpoint's union created rows the action endpoints (mcp-rest tools/list, Connect, credential writes) would then deny. Fixed in dea4092b68 one level down instead of per endpoint: the admitted-user context (the same auth a gateway session resolves with) is appended inside build_effective_auth_contexts for UI session credentials, so the page list, tool counts, Connect toggles, and credential endpoints all resolve identically, and the list endpoint's one-off union is deleted as dead code. Caller-passed virtual keys are still never widened (the is_ui_session_credential gate moved into the context builder with the same semantics) and a reload failure falls back to team contexts only. Pinned at the owner (contexts include the admitted context; caller-passed keys never consult it; reload failure falls back) and at the endpoint through the real context builder (a server granted only via user object permission is listed and flagged); the append is mutation-checked and the rest-endpoints suite passes unchanged

**Comment by tin-berri** [comment]
bugbot run

**Comment by tin-berri** [comment]
@greptileai rereview

File: litellm/proxy/_experimental/mcp_server/ui_session_utils.py [hunk 3]
```diff
@@ -68,12 +77,36 @@ async def resolve_ui_session_team_ids(
...
-    return [user_api_key_auth]
+    team_contexts = (
+        [clone_user_api_key_auth_with_team(user_api_key_auth, team_id) for team_id in resolved_team_ids]
+        if resolved_team_ids
+        else [user_api_key_auth]
+    )
+    admitted_context = await _admitted_user_context(user_api_key_auth)
+    if admitted_context is None:
+        return team_contexts
+    return [*team_contexts, admitted_context]
```
[the remark this answers is not in the document]
Comment (tin-berri) [reply litellm/proxy/_experimental/mcp_server/ui_session_utils.py:139]: Correct, and fixed in 84f525bc71, deliberately not with per-server context bookkeeping: preserving which context granted each server would be a third parallel implementation of source tracking that admitted_source_grants and _resolve_admitted_subject_tools already own. Instead the tool routes now choose one principal at the boundary: acting_user_auth (in ui_session_utils, beside the other UI-session owners) swaps a non-admin dashboard session for the admitted-subject auth, the same identity a gateway session resolves with, as the first statement of list_tool_rest_api and call_tool_rest_api. Every downstream check then binds through arms that already exist: get_allowed_mcp_servers takes the admitted branch for reachability, get_allowed_tools_for_server dispatches to the per-source fail-closed tool union as its first statement, and rate limits and billing attribution ride the same auth. Admin sessions keep their operator view unchanged and caller-passed credentials are never widened. Pinned at the owner (admitted for non-admin sessions, identity for admin sessions and passed keys, fallback on reload failure) and at the route (server resolution and the tool fetch both receive the admitted auth for a non-admin session); the route swap is mutation-checked and all 277 tests across the three mapped suites pass

**Comment by tin-berri** [comment]
@greptileai

**Comment by tin-berri** [comment]
@greptileai can you review this, give me a score and explain why

File: ui/litellm-dashboard/src/components/chat/MCPAppsPanel.tsx [hunk 4]
```diff
@@ -220,6 +235,8 @@ const MCPAppsPanel: React.FC<Props> = ({ accessToken, selectedServers, onChange,
       }
       return;
     }
+    const target = serversRef.current.find((s) => s.server_id === serverId || nameOf(s) === serverName);
+    if (target && connectUnavailabilityLabel(target) !== null) return;
```
[the remark this answers is not in the document]
Comment (tin-berri) [reply ui/litellm-dashboard/src/components/chat/MCPAppsPanel.tsx]: Correct, and fixed in e1678c572a at the class rather than at the call. Tightening the guard to `!target ||` would have blocked one symptom of a stale reference the component was still free to hold: the same copy also kept the detail view rendering a server the list had dropped, and kept its tools query alive against it. The component now stores the selected server's id and derives the row from the list (`servers.find(...)`), so a server a refetch removes cannot be the detail view's subject at all, and `handleToggle` takes the server it is toggling, which deletes the lookup that could miss. Pinned by a test that opens the detail view, refetches a list without that server, and asserts the view falls back to the grid with no connect action anywhere; storing the object again fails it

Two other things went in with it, both surfaced by reading the round 7 swap rather than reported. `admitted_user_context` is now the single owner of "what identity does this dashboard session act as", so the management endpoint's private copy of the reload, the HTTPException swallow and the logging is gone along with its dead empty-user_id branch; that owner also carries the request's tracing span onto the admitted principal, since `_reload_admitted_user` builds a fresh auth from the user row and has none of its own, so the swap had been detaching every downstream lookup and the tool-call logging from the request's trace. And toolset scoping now shares one owner with the swap on the tools list route: the admitted subject resolves per grant source and a team source deliberately carries none of the caller's `object_permission`, so a toolset narrowing layered on top would evaporate on every team-granted server, admitting the request through the user's toolset grant and then serving tools from servers the toolset never named. A request carrying a toolset name keeps the caller's own credential, exactly as before the swap

Route complexity is one below where it started, the three mapped suites are green at 279 tests, and each change is mutation-checked

**Comment by tin-berri** [comment]
@greptileai can you review, give me a score and explain why

**Comment by tin-berri** [comment]
bugbot run

**Comment by tin-berri** [comment]
Real, and fixed in c12085aea3, along with a second instance of it that was reachable by the same refetch

You are right that `handleToggle` decided at click time and committed after the await. It now re-asks at the commit, through one predicate: `connectableNow(serverId)` resolves the id against the current list and returns the row only if it is still connectable, so absence fails closed instead of reading as "nothing to block". The same predicate guards entry, so the two edges of the await cannot drift

Reading the rest of the panel's async surface for the same shape turned up the worse one. The load pipeline's cancel flag was shared across runs, and the successor's effect body reset it to `false` before the predecessor's fetch resolved, so a superseded load could still reach `setServers` and put the dropped server back on the page outright, not just into the selection. That flag is now a per-effect local that only its own cleanup can clear, which also restores unmount cancellation for the chunked tool-count loop, and the load hands its own liveness check down to the tool-count and oauth-status writes rather than having them consult a flag every other run shares

Both are pinned by tests that hold the in-flight request open across the refetch and then release it: one asserts `onChange` is never called for the dropped server, the other asserts a superseded load cannot restore it. Each mutation kills exactly its own test and no other. Panel suite is 10 green, the chat suites 45, dashboard lint clean with component complexity one below where this PR started

**Comment by tin-berri** [comment]
@greptileai

**Comment by tin-berri** [comment]
bugbot run

File: ui/litellm-dashboard/src/components/chat/MCPAppsPanel.tsx [hunk 5]
```diff
@@ -124,13 +124,30 @@ const MCPAppsPanel: React.FC<Props> = ({ accessToken, selectedServers, onChange,
...
+  const connectUnavailabilityLabel = useCallback(
+    (s: MCPServer): string | null => {
+      if (!connectMode) return null;
+      if (isUnsupportedOnGatewayConnect(s.auth_type)) return "Not supported on this connection";
+      return null;
+    },
+    [connectMode],
+  );
+
+  const connectableNow = useCallback(
```
[the remark this answers is not in the document]
Comment (tin-berri) [reply ui/litellm-dashboard/src/components/chat/MCPAppsPanel.tsx:139]: Legit, and confirmed before fixing rather than reasoned about: a test that resolves the refetch and the in-flight Connect in the same tick, with no render flushed between them, fails on c12085aea3 with `onChange` called exactly once. The earlier regression test passed only because it awaited the detail view's disappearance first, which flushed the effect and hid the window

Fixed in a6d3b1763c at the cause. The lag was the whole defect, so the mirror is gone: `commitServers` writes `serversRef` and the state together at the one point the list is ever replaced, and the ref is now never older than the last committed list. Readers that want the newest answer (`connectableNow`, the oauth auto-select effect) get it, and rendering still derives from state, so the page is unchanged. Mutation-checked: restoring the passive-effect mirror fails the new test and only that one

Scope note, since the same shape exists twice more. `selectedServersRef` and `onChangeRef` mirror props, and I left them. Their staleness is not caused by the mirror: the parent owns that list, so a callback appending to it after an await can read a pre-update value however the ref is synced, and the real fix there is a functional `onChange`, which changes this component's public API. No reachability or authorization decision reads either one, which is what made `serversRef` worth fixing here

**Comment by tin-berri** [comment]
bugbot run

**Comment by tin-berri** [comment]
@greptileai

**Review by mateo-berri — APPROVED** [review]
LGTM. Thanks!
